pkg/ring/token_navigator.go

```diff
@@ -0,0 +1,269 @@
+package ring
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+)
+
+// TokenNavigator provides utility methods for traversing the ring.
+type TokenNavigator []TokenDesc
+
+// NewTokenNavigator creates a new TokenNavigator given a r, creating a
+// copy of the token list.
+func NewTokenNavigator(r *Desc) TokenNavigator {
```

Comment:
nitpick: I would move this to *Desc type instead (`ToTokenNavigator` or `GetDenormalisedTokens`) – to avoid referencing *Desc from this file, and also to avoid passing *Desc as parameter, if it can have this as method directly. But it's a minor detail.